### DISCUSSION TITLE
About the issue of long subtitle texts without any punctuation

### DISCUSSION THREAD

**Discussion by gkngkngkn** [Purfview/whisper-standalone-win#130] [General]
Just as mentioned before in the subtitledit project, I encounter situations where there are no punctuation marks in a large portion of the text when transcribing some video files using Fast-whisper. I use large v2 model to transcribe only english video.  the only parameter I add is “**--sentence**" to keep putting out compelet sentence to translate. All other parameters are set to default.

I have tried adding intitial_prompts such like --intitial_prompt "**Hello, this is the first sentence. And this is the second one. A little pause ... and we are back.**"instructing it to add punctuation marks, but this problem still persists.

What methods or parameters need to be added to force consistent punctuation throughout the entire transcription?

**Comment by despairTK** [comment]
Try different --compute_type, --beam_size, --best_of, --initial_prompt. It is impossible for a set of parameters to be suitable for all audios. When encountering audios without punctuation marks, you can only try to adjust the parameters to achieve better results.

**Reply by gkngkngkn** [reply]
Thank you. I've attempted to change several parameters mentioned above, and the preliminary conclusion is that the initial_prompt is the most crucial factor, but the stability is still inconsistent for every file.

**Reply by despairTK** [reply]
Try --initial_prompt ",.", it is best to use the large-v2 model --compute_type to automatically select. If using --initial_prompt does not produce punctuation, disable this and keep other parameters unchanged before transcribing.

**Reply by Purfview** [reply]
> Try --initial_prompt

Similar prompt is already on by default.

**Reply by gkngkngkn** [reply]
yeah, I tried it sir, the effect improved significantly but not stable in some files, I speculate that this issue is related to the underlying logic of Whisper.

**Reply by despairTK** [reply]
Transcribing audio cannot only rely on the same parameters. Different audio languages, different speaking speeds, pause intervals and other factors will affect the final result, which requires a long period of experimentation.

**Reply by gkngkngkn** [reply]
yeah,i see sir.

**Comment by Purfview** [comment]
My advice: don't touch --intitial_prompt.
Can you share the JSON file with these issues? [made with defaults, without --intitial_prompt] . Use `-f=all` to get it.
Btw, what model do you use and do you change other settings?

**Reply by gkngkngkn** [reply]
[26-February 13, 2023 ES Tape Reading AM Session.json](https://github.com/Purfview/whisper-standalone-win/files/13535309/26-February.13.2023.ES.Tape.Reading.AM.Session.json)
parameter I use : --device cuda --compute_type float16 --language en --model large-v2 --sentence -f=all
others are all defalt setting

**Reply by gkngkngkn** [reply]
The results obtained with the default parameters mentioned above are extremely poor. Due to the absence of a large number of punctuation marks, the merged subtitles become excessively lengthy, resembling the left image below. 

Subsequently, when I added the --initial_prompt "**Hello, this is the first sentence. And this is the second one. A little pause ... and we are back.**" parameter for the same audio, the effect improved significantly, as shown on the right image below.
![bfd25dbe94bd2961a1d4d0e037e7c6b](https://github.com/Purfview/whisper-standalone-win/assets/151842080/1d628357-2e43-4fc3-860d-e084d05f00e6)

 Therefore, I tentatively concluded that the initial_prompt is one of the important factors causing punctuation issues.

However, even after adding this prompt, the stability remains inconsistent, and there are still cases where punctuation is missing in files, as illustrated in the image. 
![922882cb80663782ec4cfce87184aa9](https://github.com/Purfview/whisper-standalone-win/assets/151842080/efe64c72-cccf-4932-a231-d650f831d327)

One noteworthy detail is that in these long texts, whispers should have already segmented each complete sentence, as evident from the capitalized first word of each text segment that i marked. It is merely the absence of periods that causes them to merge into a paragraph under the "**--sentence**" parameter. 

These phenomena are indeed mysterious, and so far, I have not found **a universal method or initial_prompt** that ensures consistent output of sentences with punctuation for every audio file.

**Reply by Purfview** [reply]
But `--sentence` doesn't have anything to do with your no-punctuation issue.

**Reply by Purfview** [reply]
> The results obtained with the default parameters mentioned above are extremely poor

You cram so many complains here that I don't understand what is "poor"...

> the merged subtitles become excessively lengthy

Then why you use `--sentence` defaults to make them long? I'm confused.

**Reply by gkngkngkn** [reply]
Yes, it is precisely because of the absence of periods at the end of each subtitle segment (even though Whisper has recognized them as separate complete sentences) causing the --sentence command, which separates subtitles based on periods i guess, merges them together during subtitle merging. the promblem is not becanse of --sentence command.

**Reply by gkngkngkn** [reply]
I'm not complaining sir，my english is not very good; Faster-Whisper is relatively the most complete in terms of functionality so far.and the --sentence command is very useful for me because it allows me to translate complete sentences into a language. I'm just puzzled about the core principle behind the missing punctuation.

**Reply by Purfview** [reply]
Try `--compute_type=float16`, try `--beam_size=8` `--best_of=8`.
Often `medium` model is much better with punctuations.

I wonder if it stops punctuating after the fallbacks, to see them you would need to use `--verbose=true`. [lines with`*`]

**Reply by gkngkngkn** [reply]
thank you ,let me try it.

**Reply by gkngkngkn** [reply]
I dont see the lines with*,but it looks like better with your setup, medium model does work better with punctuations truely. thx a lot sir.

**Reply by Purfview** [reply]
> I dont see the lines with*

Do you use it directly in console or SE?

**Reply by gkngkngkn** [reply]
in console sir, and I compare medium.en and large-v2 with same parameters,medium is much better. like below. it's not because of  fallbacks. And test the result is same whether i set-bs、-bs to 5 and 8 on a same model
![59e07c7467da17766017617d3920778](https://github.com/Purfview/whisper-standalone-win/assets/151842080/062d88f2-f3e6-4cf4-b402-b49487945361)
but manual adding intiatial_prompt also matters in result, comparing to default prompt after i test.

**Reply by Purfview** [reply]
If speakers are not from UK/USA then I think that `medium` is better than `medium.en`.

> manual adding intiatial_prompt also matters in result

Test effect of `?!,.` prompt. How it compares to default or to your custom prompt.

**Reply by gkngkngkn** [reply]
Here is the comparison of three different prompts for the same file. it seems custom prompt gives a better result.
![99244aae537f4b477f509c7196e6123](https://github.com/Purfview/whisper-standalone-win/assets/151842080/f216c125-a1a6-4dcc-b531-a2f305f8f4dc)
[
![faec7cf7ea3a1013f1b412c8e01dd12](https://github.com/Purfview/whisper-standalone-win/assets/151842080/85659313-52e0-4d1a-9704-e0000d08d67e)
](url)

Additionally, I have observed that if a enough long and well-punctuated intitialprompt is provided, the results with retained punctuation in long text are more stable, especially for lengthy audio files.

This is a video from an American educator with very standard pronunciation. The parameters I used are: --device cuda --compute_type float16 --language en --model medium.en --beam_size=8 --best_of=8 --sentence

**Reply by Purfview** [reply]
What if you use default prompt and add `--prefix="Okay, so the one hour is very key."`?

Btw, how it looks with just `-prompt=None`?

**Reply by gkngkngkn** [reply]
![f10cc02c10600f75b5a751256b3557a](https://github.com/Purfview/whisper-standalone-win/assets/151842080/182e8f8c-f3d1-4cec-84a9-f3decd83d142)

**Reply by despairTK** [reply]
I actually have a rather interesting solution. Generally, there are two situations in which audio without punctuation and sentences that cannot be segmented normally occurs:

One is that there is no sound or accompaniment in the first few seconds of the audio. In this case, you need to use editing software to edit directly to the beginning of the first sentence, cut out the extra few seconds of accompaniment at the beginning, and then transcribe the processed audio.

Second, if after editing out the beginning accompaniment part, the transcription still has no punctuation and the sentences cannot be separated normally, then edit again to the beginning of the second sentence and the punctuation marks and sentence separation will be displayed normally.

I don't know what the rationale is for this either. . . But before I didn't use the --initial_prompt parameter, this was how I achieved normal punctuation and sentence segmentation.

**Reply by gkngkngkn** [reply]
My speculation is that Whisper automatically transcribes the current text based on the form of the preceding text (such as tone, format, punctuation, etc.), then uses the current text as the basis for the next paragraph, like a chain reaction of dominoes. 

In the case of long videos where exceptionally lengthy statements without punctuation occur midway, Whisper might mistakenly assume that the standard for transcription is without punctuation, leading to the omission of punctuation in the subsequent text. 

The reason why I found that a longer initial prompt yields better results in testing is likely because it consistently provides sufficient lengthy text as a basis, allowing the entire text to be stably outputted in accordance with the format.

In the situation you described, I speculate that the reason is that Whisper treats the "**no sound or accompaniment**" part as equivalent to text without punctuation, as the foundational basis for the entire transcription. This is also why an initial prompt is necessary I guess.

**Reply by despairTK** [reply]
However, if a sentence is used as the initial prompt, it will affect the correct segmentation of some sentences in the audio. For example, a word that originally appeared at the end of the previous sentence was split to the beginning of the next sentence. Although these two sentences are independent sentences, changing the position of a word may change its meaning. This is the result of testing a lot of audio, so now when I use the initial prompt, I only use two simple symbols ",." to produce an effect. If it still doesn't work, I will choose another one-- compute_type, if it doesn’t work, I will cut the beginning of the audio to a certain extent.

**Comment by Purfview** [comment]
Isn't punctuation degrading after some time with your custom prompt?

**Reply by gkngkngkn** [reply]
This situation occasionally occurs, as I mentioned, Providing a longer prompt tends to yield better results than a shorter one.
Therefore, I use a complete article as the initial prompt haha.
From the results of over 100 audio files, each averaging 1-2 hours in length, it appears that approximately 5-6 of them exhibit the phenomenon of missing punctuation marks midway through. The rest perform well. I am still exploring the underlying principles behind this occurrence.

**Reply by Purfview** [reply]
OK, then keep those with the missing punctuation files at hand. I'll give you a new tool to combat these cases. 😉

**Reply by despairTK** [reply]
English audio of more than 1 hour does have a greater chance of triggering no punctuation, but Asian languages are more likely to trigger no punctuation. . . Even if the audio is only less than ten minutes, it will be easily triggered. For example, out of 10 videos in Asian languages, 4 or more will have no punctuation marks. In order to deal with this situation before, I tried many tips, as shown in the screenshot.
I even combined several sentences from the same language. Even so, Asian languages are still much better than English without punctuation. Another point is that Asian languages do not have punctuation marks like English. For example, the first sentence has punctuation marks, the second sentence has no punctuation marks, and the third sentence has punctuation marks again. Or maybe a certain sentence suddenly has punctuation marks. Here I also provide a few screenshots for reference. If you need the corresponding json file, I can also provide it.

**Reply by Purfview** [reply]
@despairTK 
> but Asian languages are more likely to trigger no punctuation

But I by default did nothing for those Asian languages. Let's try something new...
Does Chinese vs Korean vs Japanese punctuations differ?

**Reply by despairTK** [reply]
Yes, I know this is not your fault. I just guess that this situation may be the same as the reason why some English audio transcriptions have no punctuation marks, just like this feedback from [gkngkngkn].

The punctuation marks and usage habits of these three languages are different.

Chinese punctuation marks:`，。！？`

Japanese punctuation marks:` 、。！？`

Korean punctuation marks: `,.!?`

Japanese does not use`，`during pauses, but uses `、`. Korean punctuation marks are the same as English.

**Reply by gkngkngkn** [reply]
@despairTK  bro, You can try using a article perfected with punctuation in these languages as initial prompts, as long as possible. In my attempts, this method has proven to be effective, the overall situation with punctuation is good, although some issues still arise as I metioned.

**Reply by despairTK** [reply]
@gkngkngkn
I have tried, but the effect is not good. If too many sentences are used as initial prompts in Asian languages, it will affect the sentence segmentation of Asian languages, and the unexpected situation I mentioned here will occur. #https://github.com/Purfview/whisper-standalone-win/discussions/130#discussioncomment-7742353

**Comment by Purfview** [comment]
Ok, here you go, the promised test build -> https://we.tl/t-wBSWVD8hQy

New hardcoded initial_prompt presets:
`-prompt=default` - this should be better than previous default.
`-prompt=default2` - this is maybe similar to "custom", I think maybe to make this as the default preset in the new release.
`-prompt=default_zh`
`-prompt=default_ja`

`--reprompt=2` - This will re-add initial prompt to every window. So initial prompt never disappears from the prompt.

So, I would want feedback on how new prompt defaults compares to the previous version, then check if `reprompt` helps with the long audios with the punctuation problems.

**Reply by despairTK** [reply]
Here are my test results:

`--model large-v2 --beam_size 10 --best_of 10 --compute_type=float16 --sentence --verbose=true --max_comma 1000 --max_gap 6000 --batch_recursive --reprompt=2 -prompt=default_ja`

json file：[01 ja test.zip](https://github.com/Purfview/whisper-standalone-win/files/13554986/01.ja.test.zip)

`--reprompt=2 -prompt=default_ja` After adding these two parameters, the results are significantly improved, but towards the end of the audio, the punctuation marks suddenly disappear, and the punctuation marks also suddenly disappear and then reappear in the middle stage.However, it has been improved a lot compared to the old version. I also tested that some Japanese audios within ten minutes showed good punctuation marks.

`--model large-v2  --beam_size 10 --best_of 10 --compute_type=float16 --sentence --verbose=true  --max_comma 1000 --max_gap 6000 --batch_recursive --reprompt=2  -prompt=default`

json file：[01 en test.zip](https://github.com/Purfview/whisper-standalone-win/files/13555072/01.en.test.zip)

The English audio test sample is still the same as the old version, and the output SRT file has no punctuation marks at all.

If my parameters are set incorrectly, please tell me. I will retest.

**Reply by Purfview** [reply]
> The English audio test sample is still the same as the old version, and the output SRT file has no punctuation marks at all.

Try `-prompt=default2` on it.

**Reply by Purfview** [reply]
>  After adding these two parameters, the results are significantly improved, but towards the end of the audio, the punctuation marks suddenly disappear, and the punctuation marks also suddenly disappear and then reappear in the middle stage.

There is another new param, probably not actual for the issues in this topic. You can try it anyway:

`--prompt_max=16`, `--prompt_max=32`, `--prompt_max=64`, `--prompt_max=128`

**Reply by despairTK** [reply]
`-prompt=default2`
I tested it and got exactly the same result as `-prompt=default`, without any punctuation. I also tried adding the `--prompt_max` parameter, but the result was no punctuation at all.

`--prompt_max 16 --prompt_max 32 --prompt_max 64 --prompt_max 128`
The result of testing these parameters is that punctuation marks appear and disappear, but punctuation marks appear normally at the end of the audio. Generally speaking, `--prompt_max 128` performs best in Japanese audio, with only a few lines of subtitles where there are no punctuation marks.

This is the file I got from my test:
[--prompt_max 16.zip](https://github.com/Purfview/whisper-standalone-win/files/13556033/--prompt_max.16.zip)
[--prompt_max 32.zip](https://github.com/Purfview/whisper-standalone-win/files/13556034/--prompt_max.32.zip)
[--prompt_max 64.zip](https://github.com/Purfview/whisper-standalone-win/files/13556035/--prompt_max.64.zip)
[--prompt_max 128.zip](https://github.com/Purfview/whisper-standalone-win/files/13556036/--prompt_max.128.zip)
[-prompt=default2.zip](https://github.com/Purfview/whisper-standalone-win/files/13556037/-prompt.default2.zip)

**Reply by Purfview** [reply]
How this prompt: "Hello, this is the first sentence. And this is the second one. A little pause ... and we are back."

**Reply by despairTK** [reply]
`--model large-v2 --beam_size 10 --best_of 10 --compute_type=float16 --sentence --verbose=true --max_comma 1000 --max_gap 6000 --batch_recursive --reprompt=2 --initial_prompt "Hello, this is the first sentence. And this is the second one. A little pause ... and we are back." --output_format all`

If I'm using the wrong parameter method, please let me know so I can retest. This is the result:

[ja.zip](https://github.com/Purfview/whisper-standalone-win/files/13556288/ja.zip)
[en.zip](https://github.com/Purfview/whisper-standalone-win/files/13556291/en.zip)

**Reply by Purfview** [reply]
Can u run same as last om "en" but without reprompt.

**Reply by despairTK** [reply]
`--model large-v2  --beam_size 10 --best_of 10 --compute_type=float16 --sentence --verbose=true  --max_comma 1000 --max_gap 6000 --batch_recursive --initial_prompt "Hello, this is the first sentence. And this is the second one. A little pause ... and we are back." --output_format all`

[en.zip](https://github.com/Purfview/whisper-standalone-win/files/13556393/en.zip)

**Reply by Purfview** [reply]
It's good without reprompt, so you need more suitable audio to test it.

**Reply by despairTK** [reply]
I found two audios again, both of which were over 1 hour long. Without the `--reprompt=2` parameter, punctuation marks will suddenly disappear until the end of the audio. However, I retested these two audios with the `--reprompt=2` parameter, and the results were very good, with all the punctuation marks present.

Is this sentence also used as an initial prompt when transcribing Asian languages? Do you need to add these two parameters `-prompt=default_zh -prompt=default_ja?`

**Reply by gkngkngkn** [reply]
tested -prompt=default、 default2 and max with -reprompt2 are all without punctuation like despairTK mentioned.

I initially used this sentence""Hello, this is the first sentence. And this is the second one. A little pause ... and we are back."" as a prompt. In my tests, it worked well for shorter or slower-paced audio files. However, issues arose when dealing with certain longer files.  

With adding --reprompt=2, there has been a significant improvement based on the results of current testing. but im not sure if this sentence is the best prompt for solving this issue.
 Let me continue testing further.

**Reply by Purfview** [reply]
> I retested these two audios with the --reprompt=2 parameter, and the results were very good, with all the punctuation marks present.

Nice. I wonder if making `--reprompt=2` a default is a good idea, please do more tests, if you'll not find the bad cases for it then maybe it's safe to use by default.

> Is this sentence also used as an initial prompt when transcribing Asian languages? Do you need to add these two parameters

No. All these initial prompt test defaults are just punctuation characters passed in as tokens, because when you add them as string they are not converted to things you expect.

> With adding --reprompt=2, there has been a significant improvement based on the results of current testing. but im not sure if this sentence is the best prompt for solving this issue.

Nice. I think reprompt should be ~safe, so far I didn't noticed that it would introduce hallucinations.

**Reply by despairTK** [reply]
Does this set of parameters only test English? My plan is to put audio files in three languages: English, Korean, and Japanese into one folder for batch testing.

**Reply by Purfview** [reply]
I'm thinking of making `auto` for prompt then add different default prompts depending on the language set by the `--language`.

Is making that "Hello, this....." as default for English is good idea?
Should we translate it to other common languages and make the similar default prompts? And for other languages leave current "default2"? 

> Does this set of parameters only test English?

What set of parameters? default and default2 are made out of just latin punctuation marks.

> English, Korean, and Japanese into one folder for batch testing

Ja has different punctuation so obviously it's not good idea.

**Reply by Purfview** [reply]
@despairTK Here try this as prompt for ja, made with the google translate:

`こんにちは、最初の文です。そしてこれが2つ目です。少し休憩してください...そして戻ってきます。`

**Reply by despairTK** [reply]
`--model large-v2  --beam_size 10 --best_of 10 --compute_type=float16 --sentence --verbose=true  --max_comma 1000 --max_gap 6000 --initial_prompt "Hello, this is the first sentence. And this is the second one. A little pause ... and we are back."`

When I was just testing, there was something strange happening with one of the files. The test parameters in the previous reply are used. Enable `--reprompt=2` and there will be a sudden lack of punctuation near the end of the audio. Modified to disable `--reprompt=2` and re-transcribed, the punctuation at the end returned to normal.

![QQ截图20231206001002](https://github.com/Purfview/whisper-standalone-win/assets/111871110/5aefccec-7fd7-491a-996b-eb14588f8694)

[--reprompt=2 False.zip](https://github.com/Purfview/whisper-standalone-win/files/13562504/--reprompt.2.False.zip)
[--reprompt=2 True.zip](https://github.com/Purfview/whisper-standalone-win/files/13562505/--reprompt.2.True.zip)

**Reply by Purfview** [reply]
Look at the whole general results, not at the few sentences.

Maybe smaller prompt size would help on such files, test with 32, 64 or 128 size.
Those numbers means how many tokens [similar to words] are taken from a previous transcription into the sliding prompt window, smaller - less previous context - transcription accuracy can drop a little bit. Don't forget that initial prompt is fitting in that max size, so actual real context is a bit less.

Btw, if audio and speech is very clear maybe you can disable the context prompt. In such case reprompt has interesting effect - it would make only an initial prompt as the whole prompt for every chunk. In the reference/original code it would insert an initial prompt only to the first chunk and nothing at all to the next chunks.

**Comment by despairTK** [comment]
I tested a few dozen more audios in different languages, and also read here: https://github.com/openai/openai-cookbook/blob/main/examples/Whisper_prompting_guide.ipynb

Then I feel that the sentences filled in by `--initial_prompt` should be adjusted by myself based on the transcribed audio, and there will not be a "universal" initial prompt. `--initial_prompt` The initial purpose should only be to affect some polyphonic words, professional nouns, names, and capitalization of the transcribed content.

The impact of `--initial_prompt` on punctuation and sentence segmentation may be good or bad. The good thing is that there are punctuation marks, but the bad thing is that it imitates the sentence style of the initial prompt, and then it is originally a complete sentence, but it ends up being divided into two sentences. It is also possible that the ending word of the first line of sentences appears in the second line. Start the sentence. This kind of error is usually difficult to detect. After all, the transcribed subtitles have hundreds of lines, and it is impossible to check them line by line.

This is just my personal opinion after testing, there may be better solutions. I also hope there is a good way to solve the current problem. :)

**Reply by Purfview** [reply]
So, how is that `こんにちは、最初の文です。そしてこれが2つ目です。少し休憩してください...そして戻ってきます` vs default_ja?

**Reply by despairTK** [reply]
Both are pretty average to be honest, and I think the initial prompts need to be tailored to the transcribed audio content. For example, if the audio content is a painting video, you can fill in the initial prompt like this: `This audio content is about painting, teaching you knowledge points such as line drawing, perspective, vanishing point, color, etc.`

Such initial hints can bias the transcription content toward the theme of the painting, and filling in error-prone professional terms can effectively solve the problem of selecting homonymous words when transcribing. Even though my initial prompt was English text, it would work just as well with Japanese audio. And the punctuation marks are also displayed normally.

I think the most useful feature of the beta version you provided is` --reprompt=2`. Regarding the default `--initial_prompt` parameter, I think it is most correct to leave it to the user to fill in according to their own audio content.

**Reply by Purfview** [reply]
No, it doesn't work that way like you think. It's not ChatGPT prompt. 😉

> I think it is most correct to leave it to the user to fill in according to their own audio content.

Defaults doesn't prevent that, so there is nothing to "leave".

**Reply by despairTK** [reply]
Is it possible to implement this method, for example, an hour-long audio file is divided into several audio files of no more than ten minutes through certain parameters, transcribed in batches, and finally combined to form complete subtitles. Audio transcribed within ten minutes has more accurate time stamps, normal punctuation, and more accurate transcribed content.

Of course this is my guess. . . If it's not realistic then forget about it :)

**Reply by Purfview** [reply]
If you think that prompt somehow makes transcription inaccurate then disable it.

**Reply by gkngkngkn** [reply]
@despairTK 
**“Such initial hints can bias the transcription content toward the theme of the painting, and filling in error-prone professional terms can effectively solve the problem of selecting homonymous words when transcribing.”**

Yes, providing a specialized prompt can improve accuracy in transcribing specific professional terms or ensuring correct capitalization. while I believe punctuation should be relatively consistent across different prompts. I am wondering if there is a universal prompt, at least in terms of punctuation segmentation, that can consistently ensure accurate transcription for various types of text. For example, using a passage that covers a range of standard punctuation usage as a prompt(don't need too long).
 can test using this which created by chatgpt as prompt if u have time :)

![50e031df0845242ac814a0ea38544ed](https://github.com/Purfview/whisper-standalone-win/assets/151842080/7b7c16e1-da41-4a8b-b6e5-5e0f75ba6599)

As for the segmentation errors you mentioned that may arise from adding prompts, occasional mistakes are acceptable to me compared to the issue of without punctuation. :)

**Reply by despairTK** [reply]
@gkngkngkn 
There is no universal initial prompt. . . . If you have tested various --initial_prompt, you should find that the shorter the sentence you provide, the sentences in the transcription result will also be divided into shorter ones. The longer the sentence you provide, the longer the sentences in the transcription result will be divided. .

--initial_prompt seems to me to be more like correcting typos and helping to transcribe some nouns in professional fields as its main function. The segmented sentences and punctuation marks may simply imitate the sentence style of the initial prompt, and cannot perfectly segment the sentences correctly.

So my first choice is to use the default --initial_prompt parameter, which has minimal impact on the overall results. If there are no punctuation marks when transcribing, I will cut out the audio and then re-transcribe it. Generally, only a small number of audios will be transcribed without punctuation marks.

It is possible that waiting for the release of better models in the future will solve these problems.

**Reply by Purfview** [reply]
What about prompt reset on no punctuation in a segment?

**Reply by despairTK** [reply]
Is it resetting from the beginning of the audio or just where there is no punctuation? Anyway, you can put this feature in and we can test it.

**Reply by Purfview** [reply]
> Is it resetting from the beginning of the audio or just where there is no punctuation?

No, it would reset only after no punctuation in a segment is detected.

**Comment by Purfview** [comment]
Here is test4 -> https://we.tl/t-E0WZgFf0tR

Added `--prompt_reset_on_no_end` switch, should reset prompt if there is no end of sentence in a chunk/window.
Not tested, should write a log message on verbose when triggered.

Test it on audio where it after some time starts transcribing with no punctuations. Check if it triggers and if punctuation recovers after it.

Of course it should be used with reprompt.

**Reply by despairTK** [reply]
Test parameters: `--model large-v2 --beam_size 5 --best_of 10 --compute_type=auto --batch_recursive --output_format all --sentence --reprompt=2 --initial_prompt default`

Using the`--prompt_reset_on_no_end`parameter did improve the final result, and the sentence segmentation and punctuation worked fine.

Results comparison file：[test.zip](https://github.com/Purfview/whisper-standalone-win/files/13682339/test.zip)

**Reply by Purfview** [reply]
Hmm, that's a huge impact. How many times "prompt_reset_on_no_end" was triggered?

Could you share the audio file?

**Reply by despairTK** [reply]
The audio file is too large and cannot be uploaded.

**Reply by despairTK** [reply]
Triggered once.

![QQ截图20231215155555](https://github.com/Purfview/whisper-standalone-win/assets/111871110/b2e0244d-449a-40a6-9788-3367b12eb1e2)

**Reply by Purfview** [reply]
> The audio file is too large and cannot be uploaded.

Upload it to https://wetransfer.com
Registration is not needed there.

**Reply by Purfview** [reply]
Btw, I added that custom prompt from 3 sentences as `--initial_prompt=auto`, it works only if you set `--language`, otherwise it switches to "default".

**Reply by despairTK** [reply]
https://we.tl/t-zpzHTXy7fH

**Reply by despairTK** [reply]
What do the sentences you add look like?

**Reply by Purfview** [reply]
> What do the sentences you add look like?

"Hello, this is the first sentence. And this is the second one. A little pause... and we are back."

I've translated it to top 30 languages.

Btw, `--initial_prompt=default` consist of 4 basic punctuations and it auto switches between various languages if language is set, otherwise it uses latin.

**Reply by despairTK** [reply]
In the test version you released before, the two parameters `--reprompt=2` and `--initial_prompt "Hello, this is the first sentence. And this is the second one. A little pause ... and we are back."` are sufficient. Solve the problem of no punctuation marks in English audio.

Is the `--prompt_reset_on_no_end` parameter a solution to the problem of no punctuation marks in audio in other languages?

**Reply by Purfview** [reply]
> In the test version you released before, the two parameters --reprompt=2 and --initial_prompt "Hello, this is the first sentence. And this is the second one. A little pause ... and we are back." are sufficient. Solve the problem of no punctuation marks in English audio.

That is same as `--initial_prompt=auto --language=en`.
I think to set `auto` by default in the next release.

> Is the --prompt_reset_on_no_end parameter a solution to the problem of no punctuation marks in audio in other languages?

Yes, if you set the language or your own initial prompt.
It checks for these punctuations: `s_end_chr = ['.', '。', '!', '！', '?', '？','؟']`

**Reply by despairTK** [reply]
If it only transcribes English audio, is there no difference between this beta version and the official version?

**Reply by Purfview** [reply]
> If it only transcribes English audio, is there no difference between this beta version and the official version?

I don't understand what you mean. "Difference" in what?

**Comment by Purfview** [comment]
r160.11 with new stuff is released.
Continue the feedback on them there -> https://github.com/Purfview/whisper-standalone-win/discussions/142
